src/BodyComponent.h

```diff
@@ -0,0 +1,212 @@
+		virtual void deleteComponent(Body *body) override
+		{
+			if (luaInterface)
+				luaInterface->DeregisterComponent(body);
+
+			m_components.erase(body);
+		}
+
+		// Create a new component, or return the existing one.
+		T *newComponent(const Body *body) { return &m_components[body]; }
+		// Assert that a component exists for this body and return it
+		T *get(const Body *body) { return &m_components.at(body); }
+
+	private:
+		template <typename U>
+		friend struct BodyComponentDB::Serializer;
+		template <typename U>
+		friend struct BodyComponentDB::LuaInterface;
+
+		std::map<const Body *, T> m_components;
+	};
+
+	// Type-specific serialization implementation. Delegates to the type's
+	// internal serialization methods and provides the needed glue code.
+	//
+	// The Component::LoadFromJson method will be called after the component
+	// is constructed and added to the body, and may potentially have defaults
+	// set by the owning Body before it is deserialized.
+	template <typename T>
+	struct Serializer final : public SerializerBase {
+		Serializer(std::string name, Pool<T> *pool) :
+			SerializerBase(name),
+			pool(pool)
+		{}
+		Pool<T> *pool;
+
+		virtual void toJson(const Body *body, Json &obj, Space *space) override
+		{
+			pool->get(body)->SaveToJson(obj, space);
+		}
+
+		virtual void fromJson(Body *body, const Json &obj, Space *space) override
+		{
+			auto *component = pool->newComponent(body);
+			component->LoadFromJson(obj, space);
+		}
+	};
+
+	// Returns (and creates) a type-specific pool.
+	template <typename T>
+	static Pool<T> *GetComponentType()
+	{
+		auto iter = m_componentPools.find(TypeId<T>::Get());
+		if (iter == m_componentPools.end()) {
+			auto *pool = new Pool<T>(m_componentIdx++, TypeId<T>::Get());
+			iter = m_componentPools.emplace(TypeId<T>::Get(), pool).first;
+			m_componentTypes.push_back(pool);
+		}
+
+		return static_cast<Pool<T> *>(iter->second.get());
+	}
+
+	// Returns (if present) the polymorphic interface to component associated with the given index
+	// This differs from the type-ID and is volatile between program restarts
+	static PoolBase *GetComponentType(size_t componentIndex)
+	{
+		assert(componentIndex < m_componentTypes.size());
+		return m_componentTypes[componentIndex];
+	}
+
+	// Returns (if present) the polymorphic interface to the component type associated with the given name
+	// This name can be used for serialization or to interface with Lua
+	static PoolBase *GetComponentType(const std::string &typeName)
+	{
+		auto iter = m_componentNames.find(typeName);
+		if (iter != m_componentNames.end())
+			return iter->second;
+
+		return nullptr;
+	}
+
+	// Register a component type to be queryable from Lua with body:GetComponent()
+	template <typename T>
+	static bool RegisterLuaInterface(std::string typeName)
+	{
+		Pool<T> *pool = GetComponentType<T>();
+		if (pool->luaInterface)
+			return false;
+
+		pool->luaInterface = new LuaInterface<T>(pool);
```

Comment:
But this memory is definitely not deleted anywhere?

Reply:
This is indeed not deleted anywhere - it is technically leaked at program termination but I can add a destructor to PoolBase that deletes the interface if present.

Reply:
PoolBase deletes the luaInterface if present now.